### PR TITLE
Iotssl 495 ecjpake thr

### PR DESCRIPTION
**Pull request by robertcragie**
Change summary:

* Addition of point format handling (ecjpake.c, ecjpake.h)
* Additional key export API (config.h, ssl.h, ssl_tls.c)
* Use of MBEDTLS_ECJPAKE_C #define for adding compilation for ECJPAKE. With a pared-down config.h, I noticed some were missing for ECJPAKE extension handling. Note I changed a couple to make it look consistent with the ECDH case. I am not sure if that is right but is isn't entirely clear to me what the distinction between MBEDTLS_ECJPAKE_C and MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED is. (ssl_cli.c, ssl_srv.c)
* Clean up compilation warnings (pk_parse.c)

### CODE REVIEW

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by mpg** [comment]
See line comments regarding the two defines. Rule of thumb: inside ssl files, it's `KEY_EXCHANGE_ECJPAKE_ENABLED`, outside it's `ECJPAKE_C`. Please amend 136884c accordingly. Sorry for the confusion with the other module not using optimal defines.

If you're comfortable with rewriting history, can you squash 39a60de into the previous commit, and maybe reorder commits so that the two related to point format are consecutive? Otherwise, don't worry.

Oh, and Travis helpfully reminds me you need to run scripts/generate_features.pl after adding a new define in config.sh for the export key API :) And Clang noticed a typo in the documentation of `mbedtls_ssl_conf_export_keys_cb()`: `p_export_keyS`. Damn those test scripts are picky ;)

The rest looks good to me after a quick superficial review, thanks! I'll do a more careful review on Monday hopefully.

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by miklis** [comment]
Build finished. No test results found.

File: include/mbedtls/ecjpake.h [hunk 1]
```diff
@@ -71,6 +71,7 @@ typedef struct
     const mbedtls_md_info_t *md_info;   /**< Hash to use                    */
     mbedtls_ecp_group grp;              /**< Elliptic curve                 */
     mbedtls_ecjpake_role role;          /**< Are we client or server?       */
+    int point_format;                   /**< Format for point export        */
```
Comment (mpg) [inline include/mbedtls/ecjpake.h]: This should be set to `MBEDTLS_ECP_PF_UNCOMPRESSED` by default in `mbedtls_ecjpake_init()`. I didn't check but I think this is why the tests are failing (currently this is no initialized at all).

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -1071,6 +1078,34 @@ typedef int mbedtls_ssl_ticket_write_t( void *p_ticket,
...
+ *
+ * \note            This is required for certain uses of TLS, e.g. EAP-TLS
+ *                  (RFC 5216). The key pointers are ephemeral and therefore
+ *                  must not be stored. The keys should not be copied 
+ *                  verbatim and should be used specifically for key
+ *                  derivation purposes
+ *
+ * \param p_expkey  Context for the callback
+ * \param kb        Pointer to key block 
+ * \param mk        Pointer to master key
```
Comment (mpg) [inline include/mbedtls/ssl.h]: Why not call it "master secret" which is the standard TLS terminology?

File: include/mbedtls/ssl.h [hunk 3]
```diff
@@ -1071,6 +1078,34 @@ typedef int mbedtls_ssl_ticket_write_t( void *p_ticket,
...
+ * \brief           Callback type: Export key block and master key
+ *
+ * \note            This is required for certain uses of TLS, e.g. EAP-TLS
+ *                  (RFC 5216). The key pointers are ephemeral and therefore
+ *                  must not be stored. The keys should not be copied 
+ *                  verbatim and should be used specifically for key
+ *                  derivation purposes
+ *
+ * \param p_expkey  Context for the callback
+ * \param kb        Pointer to key block
```
Comment (mpg) [inline include/mbedtls/ssl.h]: The documentation should mention the structure of this block (in which order to we find IV, MAC, cipher key).

File: library/ssl_cli.c [hunk 4]
```diff
@@ -267,7 +267,7 @@ static void ssl_write_supported_point_formats_ext( mbedtls_ssl_context *ssl,
 
     MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, adding supported_point_formats extension" ) );
 
-    *p++ = (unsigned char)( ( MBEDTLS_TLS_EXT_SUPPORTED_POINT_FORMATS >> 8 ) & 0xFF );
+    *p++ = (unsigned char)( ( MBEDTLS_TLS_EXT_SUPPORTED_POINT_FORMATS >> 8 )  & 0xFF );
```
Comment (mpg) [inline library/ssl_cli.c]: Space introduced by mistake.

**Comment by mpg** [comment]
I just made a few line comments. Most are minor except the uninitialized point format. As far as I'm concerned, this branch is good to merge as soon as the uninit issue is resolved ans the tests are passing again; the rest can be tuned after merging if necessary.

File: include/mbedtls/ssl.h (hunk 2 above)
Comment (robertcragie) [inline include/mbedtls/ssl.h]: Agree it should be "master secret"; please change

**Comment by mpg** [comment]
See branch https://github.com/ARMmbed/mbedtls/tree/iotssl-461-ecjpake-finalization which is a continuation of this one.

**Pull request closed without merging**